### PR TITLE
Property key filter API

### PR DESCRIPTION
**Pull request by DanielBallaSZTE**
Co-authored-by: Robert Fancsik <frobert@inf.u-szeged.hu>

JerryScript-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/api/jerry.c [hunk 1]
```diff
@@ -3661,6 +3661,154 @@ jerry_foreach_object_property (const jerry_value_t obj_val, /**< object value */
...
+          && (filter & JERRY_PROPERTY_FILTER_INTEGER_INDICES_AS_NUMBER))
+      {
+        ecma_deref_ecma_string (key_p);
+        key = ecma_make_uint32_value (index);
+      }
+      else
+      {
+        ecma_ref_ecma_string (key_p);
+      }
+      ecma_collection_push_back (result_p, key);
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:3771]: What happens if both an object and its prototype has the same name?

Comment (DanielBallaSZTE) [reply jerry-core/api/jerry.c]: I've added a duplicate check to avoid this event.

File: jerry-core/api/jerry.c [hunk 2]
```diff
@@ -3661,6 +3661,184 @@ jerry_foreach_object_property (const jerry_value_t obj_val, /**< object value */
...
+      }
+      else
+      {
+        ecma_ref_ecma_string (key_p);
+      }
+
+      if ((filter & JERRY_PROPERTY_FILTER_TRAVERSE_PROTOTYPE_CHAIN) && obj_iter_p != obj_p)
+      {
+        uint32_t duplicate_idx = 0;
+        for (; duplicate_idx < result_p->item_count; duplicate_idx++)
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:3776]: I would prefer a `while()` here.

File: jerry-core/api/jerry.c [hunk 3]
```diff
@@ -3661,6 +3661,184 @@ jerry_foreach_object_property (const jerry_value_t obj_val, /**< object value */
...
+        {
+          ecma_value_t value = result_p->buffer_p[duplicate_idx];
+          JERRY_ASSERT (ecma_is_value_prop_name (value) || ecma_is_value_number (value));
+          if (ecma_is_value_number (value))
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:3780]: `JERRY_UNLIKELY ()` or `!ecma_is_value_number (value)` and reverse the conditions

**Review by zherczeg — APPROVED** [review]
LGTM, only a few minor things

**Comment by DanielBallaSZTE** [comment]
@zherczeg Thank you for the review, I have applied your suggested changes.

**Review by dbatyai — APPROVED** [review]
LGTM

**Pull request merged**